lib/elixir/lib/uri.ex

```diff
@@ -279,10 +280,10 @@ defmodule URI do
       "a str%69ng"
 
   """
-  @spec encode(binary, (byte -> boolean)) :: binary
-  def encode(string, predicate \\ &char_unescaped?/1)
-      when is_binary(string) and is_function(predicate, 1) do
-    for <<char <- string>>, into: "", do: percent(char, predicate)
+  @spec encode(binary, as_boolean(char)) :: binary
+  def encode(string, fun \\ &char_unescaped?/1)
```

Comment:
I really liked the name "predicate". If you absolutely wanna stick "fun" in there, maybe `predicate_fun`?

Reply:
Well, I also like that, but I had to change it, because it doesn't accept a predicate function, but any function. That's what this PR fixes. If we want to stick to the predicate function we need to error when no boolean is returned then.

Reply:
I don't think "predicate" is necessarily strictly boolean. A predicate is a function that is used as a predicate, that is, to tell if something is true or false, which in Elixir is pretty much truthy vs falsey. You could call the condition of an "if" a predicate. Let's call it `predicate` or `predicate_fun` and keep the rest of the code like it is, without checking for strictly boolean.

Reply:
reverted